src/groupeddataframe/grouping.jl

```diff
@@ -181,6 +197,319 @@ end
+GroupedDataFrame with 6 groups based on keys: a, b
+First Group (2 rows): a = :foo, b = 2
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+⋮
+Last Group (2 rows): a = :baz, b = 1
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> collect(groupvalues(gd))
+6-element Array{NamedTuple{(:a, :b),Tuple{Symbol,Int64}},1}:
+ (a = :foo, b = 2)
+ (a = :bar, b = 1)
+ (a = :baz, b = 2)
+ (a = :foo, b = 1)
+ (a = :bar, b = 2)
+ (a = :baz, b = 1)
+
+julia> groupvalues(gd, 2)
+(a = :bar, b = 1)
+
+julia> groupvalues(gd, 2, 1)
+:bar
+
+julia> groupvalues(gd, 2, :a)
+:bar
+
+julia> groupvalues(Tuple, gd, 2)
+(:bar, 1)
+```
+"""
+groupvalues(T::Type, gd::GroupedDataFrame) = (groupvalues(T, gd, i) for i in 1:length(gd))
+groupvalues(gd::GroupedDataFrame) = groupvalues(NamedTuple, gd)
+
+groupvalues(::Type{Tuple}, gd::GroupedDataFrame, i::Integer) = Tuple(gd.parent[gd.idx[gd.starts[i]], gd.cols])
+groupvalues(::Type{NamedTuple}, gd::GroupedDataFrame, i::Integer) = NamedTuple{Tuple(groupvars(gd))}(groupvalues(Tuple, gd, i))
+groupvalues(gd::GroupedDataFrame, i::Integer) = groupvalues(NamedTuple, gd, i)
+
+groupvalues(gd::GroupedDataFrame, i::Integer, col::Integer) = gd.parent[gd.idx[gd.starts[i]], gd.cols[col]]
+groupvalues(gd::GroupedDataFrame, i::Integer, col::Symbol) = groupvalues(gd, i, _groupvar_idx(gd, col))
+
+# Eltypes of the grouping columns
+_grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
+
+
+"""
+    GroupKey
+
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
+of the corresponding grouping columns and behaves similarly to a `NamedTuple`,
+but using it to index its `GroupedDataFrame` is much more effecient than using the
+equivalent `Tuple` or `NamedTuple`.
+
+Instances of this type are returned by `keys(::GroupedDataFrame)` and are not
+meant to be constructed directly.
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{P<:GroupedDataFrame}
+    parent::P
+    idx::Int
+end
+
+function Base.show(io::IO, k::GroupKey)
+    print(io, "GroupKey")
+    show(io, NamedTuple(k))
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.names(key::GroupKey) = Tuple(groupvars(parent(key)))
+Base.keys(key::GroupKey) = names(key)
+Base.propertynames(::GroupKey) = names(key)
+Base.values(key::GroupKey) = groupvalues(Tuple, parent(key), getfield(key, :idx))
+
+Base.iterate(key::GroupKey) = iterate(key, 1)
+Base.iterate(key::GroupKey, i::Integer) = i <= length(key) ? (key[i], i + 1) : nothing
+
+Base.getindex(key::GroupKey, i::Integer) = groupvalues(parent(key), getfield(key, :idx), i)
+
+function Base.getindex(key::GroupKey, n::Symbol)
+    try
```

Comment:
Why not add appropriate logic in https://github.com/JuliaData/DataFrames.jl/pull/1908/files#diff-51878bdcc3c7eb068f85c4e853b4554dR295 (i.e. check there if `n` is a grouping variable and throw `KeyError` if it is not? - you could avoid `try`-`catch` which slows things down here by throwing `KeyError` already in `_groupvaridx`), and here just call `groupvalues`?

Reply:
Sorry, I can't tell which part the link is referring to? `_groupvar_idx` returns `nothing` when the key isn't found and `strict=false`, so that could be used instead of a try/catch. Or if you think `KeyError` is the appropriate error type in all instances this function is used, I can just change the default and use it everywhere. l didn't realize try/catch was so expensive when no exception is thrown, but it seems like it is.